### DISCUSSION TITLE
Ideas on how to adapt the inkscape shortcut manager for windows?

### DISCUSSION THREAD

**Discussion by jer-j** [sleepymalc/VSCode-LaTeX-Inkscape#21] [Q&A]
I've been using this repo as a guide for adapting this for my needs in Windows, and I've been able to achieve a pretty similar result with the exception of the shortcut manager (Thank you Peterson for [super-figure](https://marketplace.visualstudio.com/items?itemName=peterson.super-figure)!) I think I can do the same thing as Karabiner-Elements and Hammerspoon using [AutoHotKeys](https://www.autohotkey.com/docs/v2/), but Windows does not have a pasteboard so I'm not sure how I could change the style (or what I would write the changes made with the shortcuts to). It would also help me greatly if you could explain what/why  `intersect` or `has_value` is needed. A general rundown of how the manager works would be helpful too.
Thanks!

**Comment by sleepymalc** [comment]
I'm not the one who writes the `init.lua` script originally, so my interpretation might be wrong. However, I believe that `intersect` checks whether a key combination is pressed, and `has_value` is just checking which is the desired style you want. Seems to me that the script does not write in the optimized way, but nevertheless, it works for now I guess...

**Comment by jer-j** [comment]
@kiryph would you be able to comment on for `init.lua` actually changes the current active editor style in Inkscape? I understand that `create_svg_and_paste` changes the component of `style` according to the the combination of keys pressed, but how is this actually sent to inkscape. In particular, could you explain purpose of lines 111-153?

**Comment by kiryph** [comment]
@jer-myjohn 

The range from 111-148 creates svg code which should be platform-independent.

Line number 150 uses Hammerspoon (macOS tool) to fill a system pasteboard (like Cmd-C) with the svg code.

Finally, line number 152 invokes pasting like you would have pressed Cmd-v (also using Hammerspoon).

> I've been using this repo as a guide for adapting this for my needs in Windows

[Hammerspoon](https://www.hammerspoon.org/) and [Karabiner Elements](https://karabiner-elements.pqrs.org/) are mac-only apps and I do not have any experience on scripting/automation on Windows.

I actually have converted this from G. Castell https://github.com/gillescastel/inkscape-shortcut-manager written in Python for Linux to Lua for macOS (as you probably knew).

Maybe the original Python code is clearer to you

https://github.com/gillescastel/inkscape-shortcut-manager/blob/master/normal.py#L109-L214

But the crucial bits are platform dependent and you have to find a Windows expert who knows how to adapt this.


> It would also help me greatly if you could explain what/why intersect or has_value is needed.

because  I am not really a Lua expert and Lua is a smaller Language compared to Python.

Python has a builtin intersection operation which afaik Lua does not have
https://docs.python.org/3/library/stdtypes.html#set

```Python
if {'s', 'a', 'd', 'g', 'h', 'x', 'e'} & combination:
```

If you know better solutions in Lua feel free to suggest. I am happy to improve my Lua skills.

**Comment by jer-j** [comment]
@sleepymalc & @kiryph 
Could either of you explain why it wasn't possible to replace the Xlib with a similar module like [`keyboard`](https://github.com/boppreh/keyboard) or [`pynput`](https://pynput.readthedocs.io/en/latest/) to simulate keyboard events? I'm just trying to find the best way I can adapt this. I'm also going try running Gilles original script (at least the inkscape part since I'm using VS code) on WSL to see if I can get that to work, but i'm not totally sure if it would since inkscape is running on the windows side and the script would be running on the Linux side. Also, did you encounter any issues with shortcuts activating when writing text in inkscape?

**Reply by sleepymalc** [reply]
I don't have experience with either `keyboard` or `pynput` packages, so I can't say much about those. For writing text in Inkscape, it's rare to get collision since the key combining rarely appears in any regular word. But yes, I did have some falsely triggered styling changes when typing text, which will make your text become, let's say, dashed or something like that. And undo command will fix that, so not a big deal.
